python/lsst/dm/rucio/register/rucio_interface.py

```diff
@@ -0,0 +1,245 @@
+
+        meta = RubinMeta(rubin_butler=1, rubin_sidecar=metadata)
+        d = RucioDID(
+            pfn=pfn,
+            bytes=size,
+            adler32=adler32,
+            md5=md5,
+            name=name,
+            scope=self.scope,
+            meta=meta,
+        )
+
+        return d
+
+    def _add_replicas(self, bundles: list[ResourceBundle]) -> None:
+        """Call the Rucio method add_replica for a list of DIDs
+
+        Parameters
+        ----------
+        bundles: `list[ResourceBundle]`
+            A list of ResourceBundles
+        """
+        dids = [bundle.get_did() for bundle in bundles]
+        self.replica_client = ReplicaClient()
+        self.replica_client.add_replicas(rse=self.rse, files=dids)
+
+    def _add_files_to_dataset(self, did_client, dataset_id: str, dids: list[dict]) -> None:
+        """Attach a list of files specified by Rucio DIDs to a Rucio dataset.
+
+        Ignores already-attached files for idempotency.
+
+        Parameters
+        ----------
+        dataset_id: `str`
+            Logical name of the Rucio dataset.
+        dids: `list [ dict [ str, str|int ] ]`
+            List of Rucio data identifiers.
+        """
+        retries = 0
+        max_retries = 2
+        while True:
+            try:
+                did_client.add_files_to_dataset(
+                    scope=self.scope,
+                    name=dataset_id,
+                    files=dids,
+                    rse=self.rse,
+                )
+                return
+            except rucio.common.exception.FileAlreadyExists:
+                # At least one already is in the dataset.
+                # This shouldn't happen, but if it does,
+                # we have to retry each individually.
+                for did in dids:
+                    try:
+                        did_client.add_files_to_dataset(
+                            scope=self.scope,
+                            name=dataset_id,
+                            files=[did],
+                            rse=self.rse,
+                        )
+                    except rucio.common.exception.FileAlreadyExists:
+                        pass
+                return
+            except rucio.common.exception.DatabaseException:
+                retries += 1
+                if retries < max_retries:
+                    time.sleep(random.uniform(0.5, 2))
+                    continue
+                else:
+                    raise
+
+    def register_to_dataset(self, bundles) -> None:
+        """Register a list of files in Rucio.
+
+        Parameters
+        ----------
+        Bundles: `list [ ResourceBundle ]`
+            List of resource bundles
+        """
+        logger.debug("register to dataset")
+
+        did_client = DIDClient()
+
+        datasets = dict()
+        for bundle in bundles:
+            dataset_id = bundle.dataset_id
+            datasets.setdefault(dataset_id, []).append(bundle)
```

Comment:
The way this API is used, all bundles have the same `dataset_id`, so this code is unnecessary.  In the future, we may have some Butler datasets that belong to more than one Rucio Dataset, but in those cases, I think we should be able to generate distinct lists of bundles per Rucio Dataset so that this will still be unnecessary.